### PR TITLE
Explore: Adds support for new loki 'start' and 'end' params for labels endpoint

### LINKED ISSUE #16788: Support new Loki labels query
**What would you like to be added**:

Loki labels endpoint `/api/prom/labels` will change and now will accept a time range so that we can fetch labels from storage, currently it default to 6 hours if nothing is passed.

example:

```
label names -> /api/prom/labels?start=1556265056101000000&end=1556286656101000000
label values -> api/prom/label/{name}/values?start=1556265056101000000&end=1556286656101000000
```

**Why is this needed**:

This fixes an issue on Loki explore where labels would disappear after ingestion was flushed since labels where only queried from ingestors.

see: 
https://github.com/grafana/loki/issues/453
https://github.com/grafana/loki/pull/521

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Fixes an issue on Loki explore where labels would disappear after ingestion was flushed since labels were only queried from ingestors.

**Which issue(s) this PR fixes**:
Closes #16788

**Special notes for your reviewer**:
See:
grafana/loki#453
grafana/loki#521

### CODE REVIEW

**Review by hugohaggmark — COMMENTED** [review]
Wow, great work! I talked to @marefr during lunch and he has some comments about the `range` that I totally forgot about.

**Comment by marefr** [comment]
We have a sort of design flaw in Explore current when you have selected a relative time range, e.g. last 6 hours, you actually have to parse the raw time range before getting the correct absolute time range in epoch ms.

Currently when running queries, we make sure to calculate the time range in epoch ms before executing the queries:
https://github.com/grafana/grafana/blob/40f95a95c011199e7fa7f69d22768480c56680a6/public/app/features/explore/state/actions.ts#L399-L411

So I guess you have to do similar thing as in row 406-407 before being able to use the range. Does this all makes sense?

**Comment by kaydelaney** [comment]
@marefr Yep, makes sense! Will fix now.

**Comment by marefr** [comment]
@kaydelaney yes please try that out. It's related to #17629

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 1]
```diff
@@ -120,7 +123,7 @@ export default class LokiLanguageProvider extends LanguageProvider {
     return { suggestions };
   }
 
-  getLabelCompletionItems({ text, wrapperClasses, labelKey, value }: TypeaheadInput): TypeaheadOutput {
+  getLabelCompletionItems({ text, wrapperClasses, labelKey, value }: TypeaheadInput, { range }: any): TypeaheadOutput {
```
Comment (torkelo) [inline public/app/plugins/datasource/loki/language_provider.ts]: can't see an update call to this function, where is this called with range passed in?

File: public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx [hunk 2]
```diff
@@ -70,6 +70,7 @@ export interface LokiQueryFieldFormProps extends ExploreQueryFieldProps<DataSour
   syntax: any;
   logLabelOptions: any[];
   syntaxLoaded: any;
+  range: TimeRange;
```
Comment (torkelo) [inline public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: If we are to add this to the query editor interface it should be added to the base QueryEditorProps interface (and property should be named timeRange so it's named the same as in PanelProps interface)

File: public/app/plugins/datasource/loki/language_provider.ts (hunk 1 above)
Comment (kaydelaney) [reply public/app/plugins/datasource/loki/language_provider.ts]: See Line 91:
```javascript
 return this.getLabelCompletionItems.apply(this, arguments);
```
and then in `LokiQueryFieldForm.tsx`:
```javascript
    const result = datasource.languageProvider.provideCompletionItems(
      { text, value, prefix, wrapperClasses, labelKey },
      { history, range }
    );
```

**Comment by marefr** [comment]
How does this implementation feel to you @kaydelaney. Would be interesting to get some comments from the developer point of view. Seems like we're pushing and pushing the range and time zone lower and lower than thru the component-chain and that each datasource/language provider needs to implement "parsing" of time range. 

Maybe it's possible to add some behavior to base classes/components of `QueryField` or `LanguageProvider`? Or alternatively pass down a function to retrieve the parsed time range upon request, guess this makes it similar to `timeSrv`. Or have some sort of global time range, but it's tricky since you may have a left and a right explore with different time ranges.

**Comment by marefr** [comment]
Should be able to rebase and continue with this as soon as #17677 is merged.

**Review by marefr — CHANGES_REQUESTED** [review]
FYI the `timeSrv` is now updated as part of #17677 so as a first step we may not need to pass the range to metadataRequest and use `timeSrv` instead.

If you still think that passing the range to metadataRequest is the way to go: A new property `absoluteRange` was introduced in #17677 that I would like you to use instead of passing both a `TimeRange` and a `TimeZone` as props downwards.

**Review by marefr — CHANGES_REQUESTED** [review]
Some initial feedback. Still haven't tested this, but planning to do so.

File: public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx (hunk 2 above)
Comment (marefr) [reply public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: @torkelo we're using a property named `absoluteRange` of type `AbsoluteRange` now. Still feel this should go into `QueryEditorProps` interface?

Comment (torkelo) [reply public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: maybe it's not needed, the range is also available on the last query request/response

Comment (marefr) [reply public/app/plugins/datasource/loki/components/LokiQueryFieldForm.tsx]: Okay so let's move on with this and we may look further into this later.

File: public/app/plugins/datasource/loki/language_provider.test.ts [hunk 3]
```diff
@@ -128,8 +128,8 @@ describe('Request URL', () => {
     await instance.refreshLogLabels(rangeMock, true);
     const expectedUrl = '/api/prom/label';
     const expectedParams = {
-      start: rangeMock.from.valueOf() * NS_IN_MS,
-      end: rangeMock.to.valueOf() * NS_IN_MS,
+      start: rangeMock.from * NS_IN_MS,
```
Comment (marefr) [inline public/app/plugins/datasource/loki/language_provider.test.ts]: Could use the `rangeToParams` here as well and maybe in other places?

File: public/app/plugins/datasource/loki/language_provider.ts [hunk 4]
```diff
@@ -67,13 +71,13 @@ export default class LokiLanguageProvider extends LanguageProvider {
     return syntax;
   }
 
-  request = (url: string) => {
-    return this.datasource.metadataRequest(url);
+  request = (url: string, params?: any) => {
```
Comment (marefr) [inline public/app/plugins/datasource/loki/language_provider.ts]: The request function of abstract class `LanguageProvider` in types/Explore.ts should be updated with `params` argument and probably string as `url` type.

**Review by marefr — CHANGES_REQUESTED** [review]
Great fixes of my last comments :+1: Added some minor comments.

Spotted one odd thing. When going to Explore and loading Loki I see a label request ending in `?` and no start/end, see request below. Not sure if this is intended? As soon as I click on Log Labels after this requests for log labels (with start/end) and label values (with start/end) are included.

image

Reloading the whole page gives me the following requests, where one of the label requests also ends in `?` and has no start/end params. Again, not sure this is intended or not?

image

**Comment by kaydelaney** [comment]
@marefr regarding the label request with no start/end params, that's just the request being made in `testDatasource()` to verify that the data source can be connected to and that labels can be retrieved. The data from the response doesn't get used so there shouldn't be any need to pass the parameters, I think.

**Review by marefr — CHANGES_REQUESTED** [review]
Looks good :+1: Just fix the conflicts and I'll approve this

**Review by marefr — APPROVED** [review]
LGTM

**Comment by davkal** [comment]
Just tested this too. Works well 🎉

**Pull request merged**